packages/kit/src/runtime/client/client.js

```diff
@@ -144,11 +144,24 @@ export function create_client({ target, base, trailing_slash }) {
 		if (!invalidating) {
 			const url = new URL(location.href);
 
+			/** @type {import('types').Navigation} */
+			const navigation = { from: url, to: url, type: 'invalidation' };
+
+			let cancelled = false;
+			const cancellable = { ...navigation, cancel: () => (cancelled = true) };
+
+			callbacks.before_navigate.forEach((cb) => cb(cancellable));
+
 			invalidating = Promise.resolve().then(async () => {
-				await update(url, []);
+				if (!cancelled) {
+					stores.navigating.set(navigation);
+					await update(url, []);
+					callbacks.after_navigate.forEach((cb) => cb(navigation));
+				}
 
 				invalidating = null;
 				force_invalidation = false;
+				stores.navigating.set(null);
```

Comment:
Can this result in a wrong `navigation` state when the user clicks on a link while the invalidating is in progress?
Would probably only be noticeable on a slow connection.